### ISSUE TITLE
tf.linalg.diag crash (abort)

### ISSUE THREAD

**Issue by DNXie** [tensorflow/tensorflow#57713]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.11.0-dev20220914

### Custom Code

No

### OS Platform and Distribution

Ubuntu 18.04.4 LTS (x86_64)

### Python version

3.7.6

### CUDA/cuDNN version

N/A

### Current Behaviour?

tf.linalg.diag crash (abort)

Also reproduced in the [gist](https://colab.research.google.com/drive/18ixYmcwDLRbqTLbD4AHXq1Qqt74aDY5v?usp=sharing)

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
tf.linalg.diag(k=1070828000000, diagonal=np.ones((2,2,2,2)))
```

### Relevant log output

```shell
2022-09-15 19:38:46.149294: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-09-15 19:38:46.149328: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2022-09-15 19:38:46.149367: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
2022-09-15 19:38:46.149684: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-15 19:38:46.188085: F tensorflow/core/framework/tensor_shape.cc:404] Check failed: 0 <= new_num_elements (0 vs. -3186289596827017184)
Aborted (core dumped)
```
</details>

**Comment by tilakrayal** [comment]
@DNXie,
I was facing a different error stating **OOM when allocating tensor with shape[1381143300,1381143300]** while executing the given code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/b0acb10dc1dad096ca0d33b3f58ac3b3/untitled599.ipynb). Thank you!

**Comment by tilakrayal** [comment]
@DNXie,
I was able to reproduce the issue on tensorflow v2.8, v2.9 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/91728b9e9b57d3ebd102eac2462bde17/untitled604.ipynb).

**Comment by yongtang** [comment]
Created a PR #57788 for the fix.

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.11.1 (v2.11.0-94-ga3e2c692c18).

- `tf.linalg.diag`, `tf.compat.v1.linalg.diag`, `tf.compat.v1.matrix_diag`

It seems to be fixed in tensorflow 2.12.0 (v2.12.0-rc1-12-g0db597d0d75) and later versions.
